Refactor testing utilities

jarrodmillman opened the pull request:
As discussed in #4582, change `assert_edges_equal`, `assert_graphs_equal`, and `assert_nodes_equal` to be more pytest-idiomatic.  For example, `assert_edges_equal` becomes the Boolean function `edges_equal` and then the assert is done the testing file (i.e., `assert edges_equal(edges1, edges2)`).

This also makes these utility functions useful in nontesting situations where you want to compare edges, but not raise an exception based on the result.

<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

jarrodmillman commented:
@rossbar @dschult I haven't deprecated anything yet (I will do that at the end).  And I haven't changed anything in the test files yet.  This is just a refactor of each `assert_*_equal` into two functions calls.  Before I change the test files and add the deprecations, I have two questions:

1. Does the refactor look correct?  Since the asserts shortcircuit things when the value is False, I return False immediately.
2. Should the new utility functions stay here?  We've never documented them and we are already changing their names slightly.  For example, we could move these three functions to `networkx/utils/testing.py`.  Or, since as Dan pointed out they could potentially now be useful in non-testing situations, we could move them to `networkx/utils/misc.py`.

rossbar commented:
> Does the refactor look correct? Since the asserts shortcircuit things when the value is False, I return False immediately.

LGTM :+1: 

> Should the new utility functions stay here? We've never documented them and we are already changing their names slightly. 

My vote would be to leave them somewhere non-public at least for now. The main reason being that something like `graph_equality` may not have the same meaning in the context of writing tests as it might in other contexts (my understanding is that "graph equality" in particular is a thorny concept). My vote would be a two step approach 1) refactor the tests using these new fn defitions then 2) figure out if/how best to publicly expose the functions.

dschult commented:
This looks good to me too!

As far as placement, it looks like these are now the only functions in `testing/utils.py`.  So moving them to `utils.misc.py` would allow you to reduce the `testing` directory to a single module called `test.py` and no `tests` directory.  Furthermore, the distinction between `utils` and `testing/utils` has always been a little strange to me.  If/when we move them, we need to make the doc_strings quite clear about the fact that these are checking equality in a python object sense, and not in a graph theoretic sense (which is complicated). But we have had a recent request for such a feature. So I think they should eventually become public -- though not advertised heavily perhaps.  The `utils` subpackage does that by requiring the user to import is specifically (at least until lazy loading gets more common). As for timing -- I am fine if you want to do the move now or in a separate PR.  I suspect this will get pretty hairy and touch a lot of code. So, whatever makes that process better for you is good with me. :}

jarrodmillman commented:
I keep forgetting about `testings/tests.py`.  I never use it, since I use `pytest` directly.  I don't believe it is mentioned in the documentation and the testing subpackage isn't included in the documentation.  By default, `pip install networkx` doesn't install pytest.

@dschult  Should we also deprecate `run` in `testings/tests.py`?   Then if we move the testing utility functions to `utils`, we can completely remove `networkx/testing/` in 3.0.

I will move the new utility functions in this PR, add docstrings to the new utility functions, and update all the testing files to use the new utility functions.  Everything else would be deprecated and removed in 3.0.

dschult commented:
Yes, I think that we can remove the `run` functionality in favor of using pytest.  

I don't think I fully understand the implications of that change. But I have the impression it is moving from an older install ecosystem where testing was done within the python interpreter instead of via a standalone program. And it seems the pip install features and impact (it is also a standalone program) has grown to the point where the `run` function is unlikely to ever be used. Let's make that break and remove some code. :}

jarrodmillman commented:
@dschult @rossbar In 0ec4170, I moved `nodes_equal`, `edges_equal`, and `graphs_equal` from `networkx.testing.utils` to `networkx.utils.testing`.  As discussed above, this will allow us to remove `networkx.testing` in 3.0.

However, I just noticed @dschult proposed `networkx.utils.misc` instead of `networkx.utils.testing`.  I am happy w/ either and wanted to see what folks wanted before doing more.  Also, I added `from networkx.utils.testing import *` to `networkx/utils/__init__.py`.  Just checking that that is what we want to do.

Once we decide on the location, I will:
- [ ] add docstrings to  `nodes_equal`, `edges_equal`, and `graphs_equal`
- [ ] use  `nodes_equal`, `edges_equal`, and `graphs_equal` in the test files

jarrodmillman commented:
@rossbar @dschult This is done and ready for review.  In particular, it would be great if you could review my docstrings 85d4041.  Should we say more?  E.g., should we add notes to the docstring?  If so, what?

rossbar reviewed (approved):
LGTM - I think the docstrings are fine; I can't think of anything extra that's needed. A couple other suggestions related to condensing imports to a single line but looks ready to go (and a nice improvement).

jarrodmillman commented on doc/reference/utils.rst:
```diff
@@ -22,7 +22,9 @@ Helper Functions
    groups
    create_random_state
-
+   nodes_equal
+   edges_equal
+   graphs_equal
```
I am also not sure how we should "advertise" these.  I think part of the idea of this change is to indicate these functions for nontesting cases.  I am not sure how much we should recommend it for that purpose.

Maybe we could still the functions under "Testing utilities" (as above) and mention they could be used in nontesting cases in the `Notes` part of each function's docstring.  We could remove the heading in a later release, if we decide to advertise the functions for nontesting purposes more prominently.

jarrodmillman replied:
Or we could just add a sentence here mentioning that they may be useful for nontesting purposes (but still leave them under the `Testing utilities` heading.

rossbar replied:
IMO none of the options stands out as obviously better than any other, so I'm inclined to go with the one with the minimal footprint; i.e. ignore my original suggestion and leave it as-is. The docs/organization can always be updated later if users find it confusing.
